build: add static analysis targets

yuvalif opened the pull request:
adding static analysis targets:
- [x] cppcheck: http://cppcheck.sourceforge.net/. ``make cppcheck`` should generate ``cppcheck.txt`` with the error found by the tool
- [x] iwyu: https://include-what-you-use.org/. ``make iwyu`` should generate `iwyu.txt`` file with the modification suggestion
- [x] verify that iwyu run with clang and not gcc

> Notes:
> 1. In all cases the analysis is based on ``compile_commands.json``, and therefore requires ``CMAKE_EXPORT_COMPILE_COMMANDS`` to be set.
> 2. tools are not automatically installed, and require manual installations. cmake will be prniting a message in such a case
> 3. scan-build target will be added in a separate PR

badone commented:
@yuvalif I don't think we should be adding cppcheck and iwyu as additional dependencies for ceph do you?

yuvalif commented:
> @yuvalif I don't think we should be adding cppcheck and iwyu as additional dependencies for ceph do you?

@badone agree that this is not a good idea. removed from the deps installers, and added as a message in cmake

yuvalif commented on cmake/modules/FindCppCheck.cmake:
```diff
@@ -0,0 +1,30 @@
+unset(CPPCHECK_BIN CACHE)
```
at some point i wanted to test the negative case. so i uninstalled cppcheck, but since find_program cache the program, it did not detect the issue correctly unless i unset the variable first

yuvalif commented on cmake/modules/FindCppCheck.cmake:
```diff
@@ -0,0 +1,30 @@
+find_program(CPPCHECK_BIN cppcheck)
+
+if(CPPCHECK_BIN)
+    find_file(PROJECT_FILE compile_commands.json PATH .)
```
changed indentation to 2 spaces

tchaikov commented:
i am inclined to avoid adding the support of tools not tightly coupled with Ceph into cmake.

yuvalif commented:
> i am inclined to avoid adding the support of tools not tightly coupled with Ceph into cmake.

having accessible tools encourages developers to use them. if not as make targets these could be added as scripts?

tchaikov commented:
> having accessible tools encourages developers to use them. if not as make targets these could be added as scripts?

actually, i expect the output of using CppCheck and IWYU. Brad has been posting the output of them for quite a while, nothing visible happened..

since this change has been merged, could you please remove the messages like "Could not find cppcheck / IWYU"? or i will do it.

but if these helpers do not result in improvement in these area in a long period, like 4 months, i'd revert this change.

tchaikov commented:
@yuvalif

yuvalif commented:
> > having accessible tools encourages developers to use them. if not as make targets these could be added as scripts?
> 
> actually, i expect the output of using CppCheck and IWYU. Brad has been posting the output of them for quite a while, nothing visible happened..
> 
My experience with static analysis, test coverage etc. is that people usually don't go and just fix the issues. However, if the tools are accessible, they can be used to implement the "boy (or girl) scout rule". This won't happen without some guidance and effort, but IMO achievable.
In the (far?) future, this could be even enforced as part of the PR process...
> since this change has been merged, could you please remove the messages like "Could not find cppcheck / IWYU"? or i will do it.
The message is to inform that the tool is not installed on the machine (since it is not part of the "install-deps.sh" mechanism). Feel free to remove, but I thought it could be useful to notify developers that the option exist.
> 
> but if these helpers do not result in improvement in these area in a long period, like 4 months, i'd revert this change.

tchaikov commented:
@yuvalif i understand you are adding this feature out of good will. but i don't think it'd help without much more efforts as you put. if some of us would like to add this to PR process, i am all for it. but this does not imply that we need to integrate this options into cmake. more and more output from the default `cmake` call is just distracting.
